File: packages/aegis-sdk-py/salanor_aegis_ledger/verify.py
from __future__ import annotations

from pathlib import Path
from typing import Any

from salanor_aegis_ledger.canonical import event_hash_from_body
from salanor_aegis_ledger.schema import ApsValidationError, validate_event
from salanor_aegis_ledger.store import read_events


def _body_without_event_hash(event: dict[str, Any]) -> dict[str, Any]:
    chain = event["chain"]
    return {**{k: v for k, v in event.items() if k != "chain"}, "chain": {"prev_event_hash": chain["prev_event_hash"]}}


def verify(store_path: Path | str) -> dict[str, Any]:
    events = read_events(Path(store_path))
    errors: list[str] = []
    expected_prev: str | None = None

    for index, event in enumerate(events):
        try:
            validate_event(event)
        except ApsValidationError as err:
            errors.append(f"event[{index}] schema: {err}")
            continue

        if event["chain"]["prev_event_hash"] != expected_prev:
            errors.append(f"event[{index}] prev_event_hash mismatch")

        computed = event_hash_from_body(_body_without_event_hash(event))
        if computed != event["chain"]["event_hash"]:
            errors.append(f"event[{index}] event_hash mismatch")

        expected_prev = event["chain"]["event_hash"]

    return {"ok": not errors, "event_count": len(events), "errors": errors}

File: packages/aegis-ledger-sdk/src/verify.ts
import { eventHashFromBody } from "./canonical.js";
import { validateEvent } from "./schema.js";
import { readEvents } from "./store.js";
import type { ApsEvent, VerifyResult } from "./types.js";

function bodyWithoutEventHash(event: ApsEvent): Record<string, unknown> {
  const { chain, ...rest } = event;
  return {
    ...rest,
    chain: { prev_event_hash: chain.prev_event_hash },
  };
}

/**
 * Local integrity check: schema + hash chain continuity.
 */
export function verify(storePath: string): VerifyResult {
  const events = readEvents(storePath);
  const errors: string[] = [];
  let expectedPrev: string | null = null;

  for (const [index, event] of events.entries()) {
    try {
      validateEvent(event);
    } catch (err) {
      errors.push(`event[${index}] schema: ${(err as Error).message}`);
      continue;
    }

    if (event.chain.prev_event_hash !== expectedPrev) {
      errors.push(
        `event[${index}] prev_event_hash mismatch: expected ${expectedPrev ?? "null"}, got ${event.chain.prev_event_hash}`,
      );
    }

    const computed = eventHashFromBody(bodyWithoutEventHash(event));
    if (computed !== event.chain.event_hash) {
      errors.push(`event[${index}] event_hash mismatch`);
    }

    expectedPrev = event.chain.event_hash;
  }

  return { ok: errors.length === 0, event_count: events.length, errors };
}
